module/block_iterator/creator_test.go

```diff
@@ -0,0 +1,334 @@
+	blocks = append(blocks, additionalBlocks...)
+
+	// Create another iterator
+	iterator, err = creator.Create()
+	require.NoError(t, err)
+
+	// Iterate through initial blocks
+	for i := 0; i < len(additionalBlocks); i++ {
+		blockID, ok, err := iterator.Next()
+		require.NoError(t, err)
+		require.True(t, ok)
+		visitedBlocks = append(visitedBlocks, blockID)
+	}
+
+	// No more blocks to iterate
+	_, ok, err := iterator.Next()
+	require.NoError(t, err)
+	require.False(t, ok)
+
+	// Verify all additional blocks were visited exactly once
+	for i, block := range additionalBlocks {
+		require.Equal(t, block.ID(), visitedBlocks[i], "Block %v was not visited", block.Height)
+	}
+
+	// Verify no extra blocks were visited
+	require.Equal(t, len(additionalBlocks), len(visitedBlocks), "Unexpected number of blocks visited")
+
+	// Verify the final checkpoint
+	next, err = iterator.Checkpoint()
+	require.NoError(t, err)
+	require.Equal(t, uint64(9), next, "Expected next height to be 9 (last height + 1)")
+	savedHeight, err := progress.ProcessedIndex()
+	require.NoError(t, err)
+	require.Equal(t, uint64(9), savedHeight, "Expected next height to be 9 (last height + 1)")
+}
+
+func TestCanResume(t *testing.T) {
+
+	root := &flow.Header{Height: 0}
+	// Create mock blocks
+	blocks := []*flow.Header{
+		{Height: 1},
+		{Height: 2},
+		{Height: 3},
+		{Height: 4},
+		{Height: 5},
+	}
+
+	// Mock getBlockIDByHeight function
+	getBlockIDByHeight := func(height uint64) (flow.Identifier, error) {
+		for _, block := range blocks {
+			if block.Height == height {
+				return block.ID(), nil
+			}
+		}
+		return flow.Identifier{}, fmt.Errorf("block not found at height %d", height)
+	}
+
+	// Mock progress tracker
+	progress := &mockProgress{}
+
+	// Mock latest functions
+	latest := func() (*flow.Header, error) {
+		return blocks[len(blocks)-1], nil
+	}
+
+	// Create iterator
+	creator, err := NewHeightBasedCreator(
+		getBlockIDByHeight,
+		progress,
+		root,
+		latest,
+	)
+	require.NoError(t, err)
+
+	iterator, err := creator.Create()
+	require.NoError(t, err)
+
+	// Iterate through blocks
+	visitedBlocks := make([]flow.Identifier, 0, len(blocks))
+	for i := 0; i < 3; i++ { // iterate up to Height 3
+		blockID, ok, err := iterator.Next()
+		require.NoError(t, err)
+		if !ok {
+			break
+		}
+		visitedBlocks = append(visitedBlocks, blockID)
+	}
+
+	// save the progress
+	_, err = iterator.Checkpoint()
+	require.NoError(t, err)
+
+	// Additional blocks to be added later
+	additionalBlocks := []*flow.Header{
+		{Height: 6},
+		{Height: 7},
+		{Height: 8},
+	}
+
+	// Update blocks so that the latest block is updated, and getBlockIDByHeight
+	// will return the new blocks
+	blocks = append(blocks, additionalBlocks...)
+
+	// simulate a restart by creating a new creator with a different latest
+	newCreator, err := NewHeightBasedCreator(
+		getBlockIDByHeight,
+		progress,
+		root,
+		latest,
+	)
+	require.NoError(t, err)
+
+	newIterator, err := newCreator.Create()
```

Comment:
This test simulates a restart, which will create a new creator and resume from last checkpointed iterated block and iterate until the latest.